### PR TITLE
[SHARE-932][SHARE-961][Improvement] API refactor, allow updating Sources

### PR DESCRIPTION
**Pull request by aaxelb**
Combines #711 and #713, with additional cleanup.

* Reorganize API code to be more maintainable and consistent
  * Clump view/serializer/urls together for each type (e.g. `api/sources/serializers.py`, `api/sources/views.py`, etc.)
* Fork django-rest-framework-json-api, add some optional configuration to support our wants/needs
  * Allow using a custom ID in the API, rather than hardcoding to `pk`
  * Allow omitting keys with null values, configured per model
  * Allow returning data with error responses
* When trying to create a source with a name/title that already exists, return 409 and include the existing source in the response
* Allow updating sources with PUT/PATCH
  * For now, no object-level permissions, just global "can change source" permission
  * NOTE: Updating a source's `long_title` will NOT update the elastic index automatically

### CODE REVIEW

File: api/shareobjects/serializers.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from api.base.serializers import ShareSerializer
+
+
+class ShareObjectSerializer(ShareSerializer):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # remove hidden fields
```
Comment (laurenbarker) [inline api/shareobjects/serializers.py]: Unclear comment

File: api/sourceregistrations/views.py [hunk 2]
```diff
@@ -42,12 +43,3 @@ class ProviderRegistrationViewSet(viewsets.ModelViewSet):
 
     def get_queryset(self):
         return ProviderRegistration.objects.filter(submitted_by_id=self.request.user.pk)
-
-    def create(self, request, *args, **kwargs):
```
Comment (laurenbarker) [inline api/sourceregistrations/views.py]: Should this be deleted?

File: api/urls.py [hunk 3]
```diff
@@ -1,137 +1,30 @@
...
-
-model_schema_patterns = [
-    url(r'schema/{}/?'.format(v.MODEL.__name__), v.as_view())
-    for v in views.ModelSchemaView.model_views
-]
+app_name = 'api'
 
 urlpatterns = [
+    url('^$', RootView.as_view()),
+    url('^', include('api.banners.urls')),
```
Comment (laurenbarker) [inline api/urls.py]: Do you want to change these?

File: api/users/serializers.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+from rest_framework_json_api import serializers
+
+from share import models
+
+from api.base import ShareSerializer
+
+
+class ShareUserSerializer(ShareSerializer):
+    def __init__(self, *args, token=None, **kwargs):
```
Comment (laurenbarker) [inline api/users/serializers.py]: Remove `token` -Abram

File: api/users/views.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    """
+    Returns details about the currently logged in user
+    """
+    serializer_class = ShareUserSerializer
+
+    def get_queryset(self):
+        return models.ShareUser.objects.filter(pk=self.request.user.pk)
+
+
+# TODO Remove me
```
Comment (laurenbarker) [inline api/users/views.py]: This is probably related to moving things to `users/me`. We should update this comment and reference https://openscience.atlassian.net/browse/SHARE-586.

File: api/util.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+class HttpSmartResponsePermanentRedirect(http.HttpResponsePermanentRedirect):
+    status_code = 308
+
+
+def absolute_reverse(view_name, *args, **kwargs):
+    return '{}{}'.format(settings.SHARE_API_URL.rstrip('/'), reverse(view_name, *args, **kwargs))
+
+
+def validation_error_has_code(exc, code):
```
Comment (laurenbarker) [inline api/util.py]: Do we still need this?

File: share/models/ingest.py [hunk 7]
```diff
@@ -82,10 +82,13 @@ class Source(models.Model):
     canonical = models.BooleanField(default=False, db_index=True)
 
     # TODO replace with Django permissions something something, allow multiple sources per user
```
Comment (laurenbarker) [inline share/models/ingest.py]: It would be great to link a ticket here.

File: api/sourceregistrations/views.py (hunk 2 above)
Comment (aaxelb) [reply api/sourceregistrations/views.py]: Yeah, it doesn't do anything that the default `create` doesn't do

File: api/urls.py (hunk 3 above)
Comment (aaxelb) [reply api/urls.py]: I'm fine with them for now, I think.

**Pull request merged**